lib/libc/cheri/strfcap.3
.\"-
.\" Copyright (c) 1989, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\" Copyright (c) 2020 SRI International
.\"
.\" This software was developed by SRI International and the University of
.\" Cambridge Computer Laboratory (Department of Computer Science and
.\" Technology) under DARPA contract HR0011-18-C-0016 ("ECATS"), as part of the
.\" DARPA SSITH research programme.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" the American National Standards Committee X3, on Information
.\" Processing Systems.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd October 7, 2025
.Dt STRFCAP 3
.Os
.Sh NAME
.Nm strfcap
.Nd format decoded CHERI capability
.Sh LIBRARY
.Lb libc
.Sh SYNOPSIS
.In cheri/cheric.h
.Ft ssize_t
.Fo strfcap
.Fa "char * restrict buf"
.Fa "size_t maxsize"
.Fa "const char * restrict format"
.Fa "uintcap_t cap"
.Fc
.Sh DESCRIPTION
The
.Fn strfcap
function formats information from decoding
.Fa cap
into the buffer
.Fa buf
according to the string pointed to by
.Fa format .
.Pp
The
.Fa format
string consists of zero or more conversion specifications and
ordinary characters.
All ordinary characters are copied directly into the buffer.
A conversion specification consists of a percent sign
.Dq Ql %
followed by optional precision and flags and terminated by
a single character conversion specifier.
.Pp
No more than
.Fa maxsize
characters will be placed into the array.
So long as
.Fa buf
is a valid pointer to at least
.Fa maxsize
bytes of memory, it will be NUL terminated when
.Nm
returns.
.Pp
The conversion specifications are copied to the buffer after expansion
as follows:
.Bl -tag -width "xxxx"
.It Cm \&%a
is replaced by address of the capability.
.It Cm \&%A
is replaced by a textual representation of capability attributes
enclosed in parentheses.
Zero or more of the following attributes are included:
.Pp
.Bl -tag -compact -width invalid
.It invalid
Capability's tag is clear.
.It sentry
Capability is a sealed entry.
.It sealed
Capability is sealed but not a sealed entry.
.It capmode
Executable capability uses capability mode. (RISC-V only)
.El
.Pp
If no attributes are present, the field is omitted.
.It Cm \&%b
is replaced by the base address of the capability.
.It Cm \&%B
is replaced by the binary representation of the capability (excluding
the tag) in hexadecimal form.
.It Cm \&%C
if
.Fa cap
is NULL-derived this is equivalent to
.Ql %#xa .
If
.Fa cap
is not NULL-derived it is equivalent to
.Ql %#xa [%P,%#xb-%%xt]%? %A .
This is equivalent to
the
.Ql %#p
format of
.Xr printf 3 .
.It Cm \&%l
is replaced by the length of the capability.
.It Cm \&%o
is replaced by the offset of the capability.
.It Cm \&%p
is replaced by the raw permissions.
.It Cm \&%P
is replaced by an abbreviated textual representation of permissions
listing zero or more of the following characters:
.Pp
.Bl -tag -compact -width W
.It r
LOAD permission
.It w
STORE permission
.It x
EXECUTE permission
.It R
LOAD_CAP permission, when the architecture supports separate LOAD_CAP/STORE_CAP permission
.It W
STORE_CAP permission, when the architecture supports separate LOAD_CAP/STORE_CAP permission
.It C
CAP permission, when the architecture does not support separate LOAD_CAP/STORE_CAP permission
.It M
MUTABLE_LOAD capability permission
.It E
EXECUTIVE permission (Morello only)
.El
.It Cm \&%s
is replaced by the numeric type of the capability.
.It Cm \&%S
is replaced by the numeric type of the capability unless the capability
is a sentry or unsealed in which case
it is replaced by
.Ql <sentry>
or
.Ql <unsealed>
respectively.
.It Cm \&%t
is replaced by the top address of the capability.
.It Cm \&%T
is a pseudo operation causing the capability to be treated as tagged
in future expansions and is replaced by the empty string.
.It Cm \&%v
is replaced by the binary value of the tag (mnemonic: valid bit).
.It Cm \&%?
all text between this
.Ql %?
and the next
.Ql %
expression is copied to
.Fa buf
if the next
.Ql %
expression produces non-empty output.
.It Cm \&%%
is replaced by
.Ql % .
.El
.Pp
When formatting plain numbers with
.Cm a , b , l , o , p , s , t ,
and
.Cm v
the following flags may be used:
.Bl -tag -width "xxx"
.It Cm \&#
When formatting in hexadecimal,
.Ql 0x
(or
.Ql 0X
for
.Cm X
conversions)
is prepended to the output.
.It Cm \&-
When combined with a field width, padding is placed to the right of the
field.
.It Cm \&0
Zero padding.
As with the
.Xr printf 3
Cm 0
flag.
.It Cm \&x
Print the number in hexadecimal with lowercase letters.
.It Cm \&X
Print the number in hexadecimal with uppercase letters.
.El
.Pp
Additionally, an optional decimal string field with may be specified as
well as a minimum precision starting with a period
.Cm \&.
followed by a decimal digit string.
Width, precision,
.Cm # , - ,
and
.Cm 0
flags apply to the
.Cm C
when formatting a null-derived capability.
.Sh RETURN VALUES
The
.Nm
function returns the number of characters that would have been printed
if the
.Fa size
were unlimited
(not including the trailing
.Ql \e0
used to end output to strings).
This function returns a negative number if an error occurs.
.Sh SEE ALSO
.Xr printf 3
.Sh AUTHORS
This software and this manual page were
developed by SRI International and the University of Cambridge Computer
Laboratory (Department of Computer Science and Technology) under
contract
.Pq HR0011-18-C-0016
.Pq Do ECATS Dc ,
as part of the DARPA SSITH research programme.
.Sh BUGS
Not all possible erroneous input patterns are detected.


lib/libc/cheri/strfcap.c
/*-
 * SPDX-License-Identifier: BSD-2-Clause
 *
 * Copyright (c) 2021 SRI International
 *
 * This software was developed by SRI International and the University of
 * Cambridge Computer Laboratory (Department of Computer Science and
 * Technology) under DARPA contract HR0011-18-C-0016 ("ECATS"), as part of the
 * DARPA SSITH research programme.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#include <sys/param.h>

#include <cheri/cheric.h>

#include <stdbool.h>
#include <stdio.h>
#include <string.h>

static ssize_t
_strfcap(char * __restrict buf, size_t maxsize, const char * __restrict format,
    uintcap_t cap, bool tag)
{
	char tmp[(sizeof(void * __capability) * 2) + 1], fmt[9], *fmtp;
	const char *percent, *opt_start = NULL;
	char number_fmt, *orig_buf;
	size_t size = 0;
	long number;
	int width = 1, precision = 1;
	bool alt = false, right_pad = false;
	bool comma, have_attributes;
#ifdef CHERI_FLAGS_CAP_MODE
	bool capmode = false;
#endif

#define	_OUT(str, len)							\
	do {								\
		if (size < maxsize)					\
			memcpy(buf, (str), MIN((len), maxsize - size));	\
		buf += (len);						\
		size += (len);						\
	} while(0)

#define	FLUSH_OPT()							\
	do {								\
		if (opt_start != NULL) {				\
			_OUT(opt_start, percent - opt_start);		\
			opt_start = NULL;				\
		}							\
	} while (0)

#define OUT(str)							\
	do {								\
		FLUSH_OPT();						\
		_OUT((str), strlen(str));				\
	} while (0)

	orig_buf = buf;
	have_attributes = !tag || cheri_is_sealed(cap);
#ifdef CHERI_FLAGS_CAP_MODE
	if ((cheri_perms_get(cap) & CHERI_PERM_EXECUTE) != 0 &&
	    cheri_flags_get(cap) == CHERI_FLAGS_CAP_MODE) {
		capmode = true;
		have_attributes = true;
	}
#endif

	for (; *format; ++format) {
		if (*format != '%') {
			if (size < maxsize)
				*buf = *format;
			buf++;
			size++;
			continue;
		}

		number_fmt = 'd';
		percent = format;
more_spec:
		switch (*++format) {
		case '\0':
			*orig_buf = '\0';
			return (-1);
			continue;

		case '1': case '2': case '3': case '4': case '5':
		case '6': case '7': case '8': case '9':
			width = *format - '0';
			while (*++format >= '0' && *format <= '9')
				width = (width * 10) + *format - '0';
			--format;
			goto more_spec;

		case '.':
			precision = 0;
			while (*++format >= '0' && *format <= '9')
				precision = (precision * 10) + *format - '0';
			--format;
			goto more_spec;

		case '-':
			right_pad = true;
			goto more_spec;

		case 'a':
			number = cheri_address_get(cap);
			break;

		case 'A':
			if (have_attributes) {
				OUT("(");
				comma = false;
				if (!tag) {
					OUT("invalid");
					comma = true;
				}
				switch (cheri_type_get(cap)) {
				case CHERI_OTYPE_UNSEALED:
					break;
				case CHERI_OTYPE_SENTRY:
					if (comma)
						OUT(",");
					OUT("sentry");
					comma = true;
					break;
				default:
					if (comma)
						OUT(",");
					OUT("sealed");
					comma = true;
					break;
				}
#ifdef CHERI_FLAGS_CAP_MODE
				if (capmode) {
					if (comma)
						OUT(",");
					OUT("capmode");
				}
#endif
				OUT(")");
			} else
				opt_start = NULL;
			continue;

		case 'b':
			number = cheri_base_get(cap);
			break;

		case 'B':
			FLUSH_OPT();
			for (char *bytes = (char *)&cap;
			    bytes < (char *)&cap + sizeof(cap); bytes++) {
				snprintf(tmp, sizeof(tmp), "%02hhx", *bytes);
				_OUT(tmp, 2);
			}
			continue;

		case 'C': {
			size_t ret;

			FLUSH_OPT();
			if (cheri_is_null_derived(cap)) {
				alt = true;
				number_fmt = 'x';
				number = cheri_address_get(cap);
				break;
			}
			ret = _strfcap(buf,
			    maxsize > size ? maxsize - size : 0,
			    "%#xa [%P,%#xb-%#xt]%? %A", cap, tag);
			buf += ret;
			size += ret;
			continue;
		}

		case 'l':
			number = cheri_length_get(cap);
			break;

		case 'o':
			number = cheri_offset_get(cap);
			break;

		case 'p':
			number = cheri_perms_get(cap);
			break;

		case 'P':
			if (cheri_perms_get(cap) & CHERI_PERM_LOAD)
				OUT("r");
			if (cheri_perms_get(cap) & CHERI_PERM_STORE)
				OUT("w");
			if (cheri_perms_get(cap) & CHERI_PERM_EXECUTE)
				OUT("x");
#ifdef HAS_CHERI_PERM_LOAD_STORE_CAP
			if (cheri_perms_get(cap) & CHERI_PERM_LOAD_CAP)
				OUT("R");
			if (cheri_perms_get(cap) & CHERI_PERM_STORE_CAP)
				OUT("W");
#endif
#ifdef HAS_CHERI_PERM_CAP
			if (cheri_perms_get(cap) & CHERI_PERM_CAP)
				OUT("C");
#endif
#ifdef HAS_CHERI_PERM_LOAD_MUTABLE
			if (cheri_getperm(cap) & CHERI_PERM_LOAD_MUTABLE)
				OUT("l");
#endif
#ifdef HAS_CHERI_PERM_EXECUTIVE
			if (cheri_perms_get(cap) & CHERI_PERM_EXECUTIVE)
				OUT("E");
#endif
			continue;

		case 's':
			number = cheri_type_get(cap);
			break;

		case 'S':
			switch (cheri_type_get(cap)) {
			case CHERI_OTYPE_UNSEALED:
				OUT("<unsealed>");
				continue;
			case CHERI_OTYPE_SENTRY:
				OUT("<sentry>");
				continue;
			}
			number = cheri_type_get(cap);
			break;

		case 't':
			number = cheri_top_get(cap);
			break;

		case 'T':
			tag = true;
			continue;

		case 'v':
			number = cheri_tag_get(cap);
			break;

		case 'x':
		case 'X':
			number_fmt = *format;
			goto more_spec;

		case '?':
			opt_start = format + 1;
			while(*(format + 1) != '\0' && *(format + 1) != '%')
				format++;
			if (opt_start == format + 1)
				opt_start = NULL;	/* or error? */
			continue;

		case '%':
			OUT("%");
			continue;

		case '#':
			alt = true;
			goto more_spec;
		}

		/* If we're here, we're rendering a number. */
		fmtp = fmt;
		*fmtp++ = '%';
		if (alt)
			*fmtp++ = '#';
		if (right_pad)
			*fmtp++ = '-';
		*fmtp++ = '*';
		*fmtp++ = '.';
		*fmtp++ = '*';
		*fmtp++ = 'l';
		*fmtp++ = number_fmt;
		*fmtp = '\0';
		snprintf(tmp, sizeof(tmp), fmt, width, precision, number);
		OUT(tmp);
	}

	orig_buf[MIN(size, maxsize - 1)] = '\0';
	return (size);
}

ssize_t
strfcap(char * __restrict buf, size_t maxsize, const char * __restrict format,
    uintcap_t cap)
{
	return (_strfcap(buf, maxsize, format, cap, cheri_tag_get(cap)));
}
